File: solutions/typescript/src/VendingMachine/VendingMachineState/IdleState.ts
import VendingMachine from "../VendingMachine";
import VendingMachineState from "./VendingMachineState";

export default class IdleState implements VendingMachineState {
  private machine: VendingMachine;

  constructor(machine: VendingMachine) {
    this.machine = machine;
  }

  selectItem(code: string): void {
    if (!this.machine.getInventory().isAvailable(code)) {
      console.log("Item not available.");
      return;
    }

    this.machine.setSelectedItemCode(code);
    this.machine.setState(this.machine.getItemSelectedState());
    console.log("Item Selected = ", code);
  }

  insertCoin(): void {
    console.log("Please select the product first");
  }

  dispense(): void {
    console.log("Please select the product first");
  }

  refund(): void {
    console.log("Please select the product first");
  }
}

File: solutions/java/src/vendingmachine/state/IdleState.java
package vendingmachine.state;

import vendingmachine.enums.Coin;
import vendingmachine.VendingMachine;

public class IdleState extends VendingMachineState {
    public IdleState(VendingMachine machine) {
        super(machine);
    }

    @Override
    public void insertCoin(Coin coin) {
        System.out.println("Please select an item before inserting money.");
    }

    @Override
    public void selectItem(String code) {
        if (!machine.getInventory().isAvailable(code)) {
            System.out.println("Item not available.");
            return;
        }
        machine.setSelectedItemCode(code);
        machine.setState(new ItemSelectedState(machine));
        System.out.println("Item selected: " + code);
    }

    @Override
    public void dispense() {
        System.out.println("No item selected.");
    }

    @Override
    public void refund() {
        System.out.println("No money to refund.");
    }
}
